katdal/applycal.py

```diff
@@ -0,0 +1,188 @@
+###############################################################################
+# Copyright (c) 2018, National Research Foundation (Square Kilometre Array)
+#
+# Licensed under the BSD 3-Clause License (the "License"); you may not use
+# this file except in compliance with the License. You may obtain a copy
+# of the License at
+#
+#   https://opensource.org/licenses/BSD-3-Clause
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+###############################################################################
+
+"""Utilities for applying calibration solutions to visibilities and weights."""
+
+from functools import partial
+
+import numpy as np
+import dask.array as da
+
+from katdal.categorical import CategoricalData
+
+
+def calc_delay_correction(cache, product, index, freqs):
+    """Calculate correction sensor from delay calibration solutions.
+
+    This extracts the cal solution sensor associated with `product` from
+    `cache`, then extracts the delay time series of the input specified by
+    `index` (in the form (pol, ant)) and builds a categorical sensor for
+    the corresponding complex correction terms (channelised by `freqs`).
+
+    Invalid delays (NaNs) are replaced by zeros, since bandpass calibration
+    still has a shot at fixing any residual delay. An invalid `product` results
+    in a :exc:`KeyError`.
+    """
+    all_delays = cache.get('cal_product_' + product)
```

Comment:
Is it worth parametrising the name of the cal "stream" at this point? Or are you confident that we will be limited to a single cal called "cal" for the long term?

Reply:
I'm not confident, no, but rather pragmatic (or is that lazy?) :-)

I can have a look at your cal parametrisation examples and see if it is worth it.

Reply:
Laziness is a virtue :-). Don't let my perfectionism derail you from getting something out the door.

Reply:
Left until later.